### ISSUE TITLE
Implement concise gen-matching / cross-cleaning in JaggedCandidateArray

### ISSUE THREAD

**Issue by lgray** [scikit-hep/coffea#18]
Not being addressed in the current PR want to close out a new release.

**Comment by lgray** [comment]
@areinsvo @mcremone if you wanna try your hand at this, please give it a go.

**Comment by mcremone** [comment]
Just to help myself remembering, this is the description from Jim on how to do gen-reco matching with awkward arrays:

 I'd like to show what a combinatoric problem like gen-reco matching involves. Something as common as this should probably get wrapped up as a user-callable function, but here's what it looks like in raw array manipulation.

First, we need some particles to match. I made an incredibly small sample to demonstrate. This is a good technique for getting some machinery working before scaling up to real datasets. On these contrived examples, we know what the answer should be.
gen_pt   = awkward.fromiter([[10.0, 20.0, 30.0],       [],     [40.0, 50.0]])
reco_pt  = awkward.fromiter([[20.2, 10.1, 30.3, 50.5], [50.5], [50.5]])

gen_eta  = awkward.fromiter([[-3.0, -2.0, 2.0],        [],     [-1.0, 1.0]])
reco_eta = awkward.fromiter([[-2.2, -3.3, 2.2, 0.0],   [0.0],  [1.1]])

gen_phi  = awkward.fromiter([[-1.5,  0.0, 1.5],        [],     [0.78, -0.78]])
reco_phi = awkward.fromiter([[ 0.1, -1.4, 1.4, 0.78],  [0.78], [-0.77]])

gen  = uproot_methods.TLorentzVectorArray.from_ptetaphim(gen_pt, gen_eta, gen_phi, 0.2)
reco = uproot_methods.TLorentzVectorArray.from_ptetaphim(reco_pt, reco_eta, reco_phi, 0.2)
The gen and reco are now jagged arrays of TLorentzVector. I set this up so that the first event has one-to-one matches with an extra reco particle and the order is wrong, the second event is empty, and the third event has two gen particles for one reco.
gen
[[TLorentzVector(0.70737, -9.9749, -100.18, 100.68) TLorentzVector(20, 0, -72.537, 75.244) TLorentzVector(2.1221, 29.925, 108.81, 112.87)] [] [TLorentzVector(28.437, 28.131, -47.008, 61.724) TLorentzVector(35.546, -35.164, 58.76, 77.154)]]
reco
[[TLorentzVector(20.099, 2.0166, -90.034, 92.272) TLorentzVector(1.7167, -9.953, -136.73, 137.11) TLorentzVector(5.15, 29.859, 135.05, 138.41) TLorentzVector(35.901, 35.516, 0, 50.5)] [TLorentzVector(35.901, 35.516, 0, 50.5)] [TLorentzVector(36.254, -35.155, 67.45, 84.26)]]
We could cross-join them to get all combinations of gen and reco in the same event, but "cross" without any arguments puts all combinations at the same level of jaggedness, making it hard to do something asymmetric with them, like find the best reco for a given gen.
gen.cross(reco)
[[(TLorentzVector(0.70737, -9.9749, -100.18, 100.68), TLorentzVector(20.099, 2.0166, -90.034, 92.272)) (TLorentzVector(0.70737, -9.9749, -100.18, 100.68), TLorentzVector(1.7167, -9.953, -136.73, 137.11)) (TLorentzVector(0.70737, -9.9749, -100.18, 100.68), TLorentzVector(5.15, 29.859, 135.05, 138.41)) ... (TLorentzVector(2.1221, 29.925, 108.81, 112.87), TLorentzVector(1.7167, -9.953, -136.73, 137.11)) (TLorentzVector(2.1221, 29.925, 108.81, 112.87), TLorentzVector(5.15, 29.859, 135.05, 138.41)) (TLorentzVector(2.1221, 29.925, 108.81, 112.87), TLorentzVector(35.901, 35.516, 0, 50.5))] [] [(TLorentzVector(28.437, 28.131, -47.008, 61.724), TLorentzVector(36.254, -35.155, 67.45, 84.26)) (TLorentzVector(35.546, -35.164, 58.76, 77.154), TLorentzVector(36.254, -35.155, 67.45, 84.26))]]
So "cross" now has a new option: gen.cross(reco, nested=True) increases the jaggedness by one so that all combinations with the same gen particle are in the same list. Instead of [(g1, r1), (g1, r2), (g2, r1), (g2, r2)] in an event, we have [[(g1, r1), (g1, r2)], [(g2, r1), (g2, r2)]].
pairing = gen.cross(reco, nested=True)
[[[(TLorentzVector(0.70737, -9.9749, -100.18, 100.68), TLorentzVector(20.099, 2.0166, -90.034, 92.272)) (TLorentzVector(0.70737, -9.9749, -100.18, 100.68), TLorentzVector(1.7167, -9.953, -136.73, 137.11)) (TLorentzVector(0.70737, -9.9749, -100.18, 100.68), TLorentzVector(5.15, 29.859, 135.05, 138.41)) (TLorentzVector(0.70737, -9.9749, -100.18, 100.68), TLorentzVector(35.901, 35.516, 0, 50.5))] [(TLorentzVector(20, 0, -72.537, 75.244), TLorentzVector(20.099, 2.0166, -90.034, 92.272)) (TLorentzVector(20, 0, -72.537, 75.244), TLorentzVector(1.7167, -9.953, -136.73, 137.11)) (TLorentzVector(20, 0, -72.537, 75.244), TLorentzVector(5.15, 29.859, 135.05, 138.41)) (TLorentzVector(20, 0, -72.537, 75.244), TLorentzVector(35.901, 35.516, 0, 50.5))] [(TLorentzVector(2.1221, 29.925, 108.81, 112.87), TLorentzVector(20.099, 2.0166, -90.034, 92.272)) (TLorentzVector(2.1221, 29.925, 108.81, 112.87), TLorentzVector(1.7167, -9.953, -136.73, 137.11)) (TLorentzVector(2.1221, 29.925, 108.81, 112.87), TLorentzVector(5.15, 29.859, 135.05, 138.41)) (TLorentzVector(2.1221, 29.925, 108.81, 112.87), TLorentzVector(35.901, 35.516, 0, 50.5))]] [] [[(TLorentzVector(28.437, 28.131, -47.008, 61.724), TLorentzVector(36.254, -35.155, 67.45, 84.26))] [(TLorentzVector(35.546, -35.164, 58.76, 77.154), TLorentzVector(36.254, -35.155, 67.45, 84.26))]]]
As before, we can make function calls at the top level that apply at all levels of depth, keeping the nested array structure. pairing.i0 picks the left of each pairing (the gen particle) and pairing.i1 picks the right of each pairing (the reco particle). We can do gen.delta_r(reco) to get delta_r for all pairs in this structure. This is what we'll use as a "goodness of match" metric.
metric = pairing.i0.delta_r(pairing.i1)
[[[1.7888543819998317 0.31622776601683783 5.953990258641679 3.76807643234582] [0.22360679774997916 1.91049731745428 4.427188724235731 2.146718425877041] [4.427188724235731 6.041522986797286 0.22360679774997916 2.1256528408938276]] [] [[2.6100766272276377] [0.1004987562112092]]]
...which we want to minimize. As a reminder, argmin gives you the local index of the value that minimizes a list as a length-one list or an empty list if empty. Like all reducers (now), it applies at the deepest level of jaggedness.
index_of_minimized = metric.argmin()
[[[1] [0] [2]] [] [[0] [0]]]
This is the right structure to pass an array to select such things by "fancy indexing". Again, you get a length-one list for the minimum and an empty list if the original was empty.
metric[index_of_minimized]
[[[0.31622776601683783] [0.22360679774997916] [0.22360679774997916]] [] [[2.6100766272276377] [0.1004987562112092]]]
Now we can use this to make selections. Comparison operators turn numbers into booleans but maintain structure (as usual). This will be our mask.
passes_cut = (metric[index_of_minimized] < 0.5)
[[[True] [True] [True]] [] [[False] [True]]]
Now, to get the pairing that not only minimizes the metric but also passes the cut, we do fancy indexing and masking: first the fancy index puts the pairings into the same jagged structure (sizes of lists, all the way down in nesting), then the mask matches that structure to say what to keep and what to drop.
best_pairings_that_pass_cut = pairing[index_of_minimized][passes_cut]
[[[(TLorentzVector(0.70737, -9.9749, -100.18, 100.68), TLorentzVector(1.7167, -9.953, -136.73, 137.11))] [(TLorentzVector(20, 0, -72.537, 75.244), TLorentzVector(20.099, 2.0166, -90.034, 92.272))] [(TLorentzVector(2.1221, 29.925, 108.81, 112.87), TLorentzVector(5.15, 29.859, 135.05, 138.41))]] [] [[] [(TLorentzVector(35.546, -35.164, 58.76, 77.154), TLorentzVector(36.254, -35.155, 67.45, 84.26))]]]
Great! We have the best pairings that pass the cut! That's the answer, except that each gen-reco match is in a length-one list and each failed cut is a length-zero list (same number of lists as we had gen particles, but some gen particles couldn't be matched). Unless we're doing an efficiency study, we no longer care about that information; we need to flatten it. Careful: the list we want to flatten is not axis=0 (events); it's axis=1 (particles). We want a flat list in each event, one for each successful gen-reco match.
genrecos = best_pairings_that_pass_cut.flatten(axis=1)
[[(TLorentzVector(0.70737, -9.9749, -100.18, 100.68), TLorentzVector(1.7167, -9.953, -136.73, 137.11)) (TLorentzVector(20, 0, -72.537, 75.244), TLorentzVector(20.099, 2.0166, -90.034, 92.272)) (TLorentzVector(2.1221, 29.925, 108.81, 112.87), TLorentzVector(5.15, 29.859, 135.05, 138.41))] [] [(TLorentzVector(35.546, -35.164, 58.76, 77.154), TLorentzVector(36.254, -35.155, 67.45, 84.26))]]
The resulting list of pairs is at most as long as the list of gen particles. If we wanted to tag gen particles, we'd have to do that before flattening (where there is exactly one zero-or-one-length list per gen particle).
genrecos.counts
[3 0 1]
gen.counts
[3 0 2]
And you can see that these are the right matches. The above procedure swapped the order of recos to match the gens, dropped the spurious reco in the first event, and had no match to the first of two gens in the third event, so we only see the one that had a match.
genrecos.i0.pt
[[10.0 20.0 30.0] [] [49.99999999999999]]
genrecos.i1.pt
[[10.1 20.2 30.3] [] [50.49999999999999]]
Procedures like this have to be done interactively. You really can't work out the indexing without seeing immediate feedback. This is why Jupyter is so popular, and at some point a sample capable of sub-sample responses is needed. It might even have to be mocked, as in my example above, to have nice round numbers with easy-to-spot differences. Once the ballet of whirling razor blades is set up, however, it can move to a script and run fast.

(Note: for badly matched indexing, you get errors, not wrong answers. If you mock the data, you want uneven sizes of things to be sure you don't have something that accidentally matches indexes.)

**Comment by areinsvo** [comment]
First attempt at this is found in the GenJetMatching branch (sorry for not following the branch naming conventions) 
https://github.com/CoffeaTeam/fnal-column-analysis-tools/blob/GenJetMatching/fnal_column_analysis_tools/analysis_objects/JaggedCandidateArray.py

A few comments: 

- There are two functions, match() and argmatch(). Call them with gen.match(reco,0.3) or gen.argmatch(reco), where gen and reco are JaggedCandidateArrays. The match() function will return a mask stating whether any reco particle matched with the gen particle within the (required) deltaR parameter. The argmatch() function will return the indices of the reco particles that matched the gen particles the best. 
- For match() there is an optional parameter to also require the percent pt difference to be within a given threshold. Note that which particle's pt is in the denominator depends on whether you do gen.match(reco,0.3,0.1) or reco.match(gen,0.3,0.1). Let me know if the order I chose needs to be flipped.
- In match() there is a check to make sure the pt of the particles in the denominator is nonzero, otherwise an exception is raised. I could expand this to check for pt =0 values for both sets of objects. 
-In argmatch() there is an optional parameter to also apply a delta R cutoff.
- @lgray mentioned that it might be nice if there could be a user-defined match criteria. We could add this, or we could rely on the user to first do argmatch() and then perform any additional cuts.
- I also implemented a brute force test. Let me know if there is a more sophisticated way to do it.
